### ISSUE TITLE
Excel Custom Functions: API Calls Timing Out After 60 Seconds

### ISSUE THREAD

**Issue by dho-caissa** [OfficeDev/office-js#1184]
## Current Behavior
The excel custom functions created as part of Add-In seem to have 60 seconds request limit on API call. When API call takes more than 60 seconds, it return "Network Request Failed" error. We are  experiencing this 60 seconds request time limit in our current production Excel Add-In.  Is there way to change it in Excel or Excel Add-in?

## Steps to Reproduce, or Live Example
To demonstrate the 60 seconds timeout behavior, we created a demo project with two custom functions. One function calls a data service that takes 70 seconds to complete. The other function calls a data service that takes 5 seconds to complete. The first function failed with "Network Request Failed" error. The code for demo project can be found here. The readm.md has instructions how to run it.
https://github.com/dho-caissa/ExcelAddinDemo.
The excel custom function seems have a timeout limit on fetch() call.

Demo code to show the behavior:
` try {

        const path = 'https://localhost:44360/data/GetDataFailed';

        const response = await fetch(path);

        const text = await response.text()

        const json = JSON.parse(text);

        

        if (json === null || json === undefined) {

            return "#no data";

        } else {

            return json;

        }

    } catch (err) {

        return "#exception:" + err;

    }
;`

## Your Environment

* Platform [PC desktop, Mac, iOS, Office Online]: PC
* Host [Excel, Word, PowerPoint, etc.]: Excel

**Comment by xiaochunms** [comment]
Thanks for the effort for the demo project. I tried the demo project, but I hit the error in the very first step, i.e. Run "NPM install and Yarn install". Any suggestion to unblock here and I can move forward then?

![image](https://user-images.githubusercontent.com/1261601/82997849-80a37f00-a039-11ea-9107-954158662b0a.png)

**Comment by dho-caissa** [comment]
hi @xiaochunms, try run "NPM install" first. Then, run "yarn install". thanks

**Comment by xiaochunms** [comment]
@dho-caissa I got following error when run "yarn install". :(
![image](https://user-images.githubusercontent.com/1261601/83154825-2b4b9880-a133-11ea-9dd2-459be766b092.png)

**Comment by dho-caissa** [comment]
@xiaochunms I think NPM install should be enough to get the javascript packages. You don't need to run the "yarn install" if you don't have yarn installed. 
The main code to demo the behavior is here. The code simply calls the API and it is relatively straight forward. thanks again.
https://github.com/dho-caissa/ExcelAddinDemo/blob/master/src/Caissa.ExcelAddin/CustomFunctions/src/functions.ts
"function getFundMarketValueFail" calls api that takes more than 60 seconds. It fails with "Network Request Failed" error.

**Comment by dho-caissa** [comment]
@xiaochunms is there any update? Any suggestion? Thanks

**Comment by dho-caissa** [comment]
Hi, Is there any update on the 60 seconds timeout rule? We tested Fetch and XmlHttpRequest with the Microsoft sample project (link below). When API takes more than 60 seconds, both methods call failed. thanks  
https://docs.microsoft.com/en-us/office/dev/add-ins/excel/custom-functions-web-reqs

**Comment by pbl2108** [comment]
@xiaochunms, @RuoyingLiang do you have any update on this? It is really causing us problems.

**Comment by RuoyingLiang** [comment]
The issue is under investigating. @shaofengzhu , could you help to update the thread once there are more information?

**Comment by pbl2108** [comment]
@shaofengzhu could you please let us know if there is an update on this?

**Comment by zlatko-michailov** [comment]
Thanks for adopting custom functions! I am sorry to see you've had an unresolved issue for such a long time.

Before I investigate any further, please take the following steps, and let me know of the outcome:
1. In Excel, click on Insert > Get Add-ins.
2. Select the Store tab, if you are not already there.
3. Enter __WA104381720__ in the Search box, hit Enter, and click the Add buttons as necessary to install this add-in.
4. Back to the grid, enter these formulas in any 4 cells:
- =MICROSOFT.OFFICE.TEST.UPGRADE.CONST_DELAYED(120000)
- =MICROSOFT.OFFICE.TEST.UPGRADE.CONST_DELAYED(75000)
- =MICROSOFT.OFFICE.TEST.UPGRADE.CONST_DELAYED(60000)
- =MICROSOFT.OFFICE.TEST.UPGRADE.CONST_DELAYED(30000)

The argument is the number of milliseconds the function will delay the reporting of its result. So you'll have to wait for at least 2 minutes for all of them to finish.

Please report here what happens after 2 minutes.

Thanks,
Zlatko

**Comment by JunielKatarn** [comment]
This issue has been identified as a hard-coded timeout in our JavaScript HTTP client.
Work in progress to increase this limit, or make it configurable.

**Comment by pbl2108** [comment]
@zlatko-michailov (мерси много!) thank you so much for looking into this!

I ran the functions with the 4 parameters you suggested and they all returned 42 as shown on the screenshot. I am not sure if they are making HTTP calls to a server, or the specified delay is completely happening on the client, because the issue we are experiencing with functions that make HTTP calls that take more than 60 seconds.

![image](https://user-images.githubusercontent.com/2523333/89604268-eb8ee480-d838-11ea-8e83-a2643ef28bbd.png)

**Comment by zlatko-michailov** [comment]
@pbl2108, it's my pleasure to help you adopt custom functions.

My add-in is very simple. It only leverages the core custom functions pipeline. I wanted to prove that the pipeline has no time limit.

@JunielKatarn is from the team that owns the networking API on top of the ReactNative runtime. He'll take it from here.

Alternatively, you can also try a "shared runtime" which uses a real browser. See if it has the same issue. All you have to change is add an element to your manifest as described here: https://docs.microsoft.com/en-us/office/dev/add-ins/reference/manifest/runtime. You can see an example here: https://github.com/OfficeDev/custom-functions/blob/master/addins/upgrade/upgrade_shared.xml

I'll keep monitoring this thread in case something related to the pipeline comes up.

**Comment by pbl2108** [comment]
Thanks again @zlatko-michailov! 

We tried the shared runtime at some point and we were not able to get it working successfully. I am not sure if the issue we ran into was the same limitation of 60 seconds, or something else. We will try again.

Excel functions defined with the JavaScript API are very powerful, and we have been using them since the early versions when they became available.

@JunielKatarn please keep us posted on the progress of removing the 60s limit or making it configurable, as this is the preferred solution for us.

**Comment by JunielKatarn** [comment]
@pbl2108 I have developed and submitted a fix in our JS runtime.
It should be available in the next monthly Office update.

The timeout will be settable using [XMLHttpRequest.timeout](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/timeout).

**Comment by pbl2108** [comment]
That's great news @JunielKatarn! Thanks for addressing the issue. We will be on the lookout for the update.

**Comment by oabdelkarim** [comment]
Any updates on this!?

**Comment by JunielKatarn** [comment]
This issue was fixed over a year ago. Feel free to close.

**Comment by sboudouk** [comment]
It is not fixed at all.

Using either `fetch` or `XMLHttpRequest` which both have default timeout set to 0 (no timeout) will timeout within 60 seconds. Excel is doing something under the hood that create this behaviour.

I don't know if it is intended or not but judging by maintainers comments I'd say no.

The consistent and simple workaround is to set manually your HTTP Client Timeout to something higher than 60 seconds and **NOT** to 0.

Simply check [XMLHttpRequest timeout documentation](https://developer.mozilla.org/fr/docs/Web/API/XMLHttpRequest/timeout)

A similar behaviour is possible with `fetch` api, just google it.


It would be great, even if it's unintended, to have it documented somewhere in Office add-in docs.

Thanks for your work.

**Comment by JunielKatarn** [comment]
@sboudouk can you please share the Office/Excel version you are using?

See https://support.microsoft.com/en-us/office/about-office-what-version-of-office-am-i-using-932788b8-a3ce-44bf-bb09-e334518b8b19

**Comment by sboudouk** [comment]
Sure. 

Excel for Mac with a 365 subscription, version 16.64 (22081401)

Similar behaviour can be observed on latest versions of Windows / Office 365 Web.

To reproduce, just setup any server that answer after 60s + and request it from a shared runtime custom function through `fetch` or `XMLHtttpRequest`. I can share my very basic Express server that I use for test if needed.

**Comment by JunielKatarn** [comment]
> Excel for Mac with a 365 subscription, version 16.64 (22081401)

 I drove the Windows implementation only.
Office uses React Native for its JS framework.
The macOS version of the HTTP module is completely separate from the Windows one.

I'll reach out to the team owning the macOS variant.

If you also notice this on Windows, would you kindly also obtain that version number?
I'll try to reproduce this behavior on my end.

**Comment by sboudouk** [comment]
Same issue on Windows, Microsoft 365 MSO Version 2207 Build 16.015427.20182 64-bit

**Comment by JunielKatarn** [comment]
> Same issue on Windows, Microsoft 365 MSO Version 2207 Build 16.015427.20182 64-bit

This version contains the fix that makes the low-level HTTP client timeout configurable, with a value of `0` meaning not timeout.
But, we have found a hard-coded 60 second timeout for CORS preflight requests (not the actual final request).
However, this would not behave as you described (the manual timeout would not override this value):
> The consistent and simple workaround is to set manually your HTTP Client Timeout to something higher than 60 seconds and NOT to 0.

We verified the lower-level components' timeout behavior, and setting it to `0` does succeed for regular (**non-preflight**) requests that take longer than 60 seconds.

Follow-up actions on our side:
- Remove the 60 second timeout from CORS preflights, unless they are expected by design.
- Contact authors of the code base used by Excel and find out whether there is an enforcement for 60 second timeouts (@shaofengzhu does your team have any insights about this?).

**Comment by sboudouk** [comment]
Thanks, great actions. I'm sure this will help the next ones facing this issue.

**Comment by guptamayank9827** [comment]
Facing the same issue on _Microsoft for Excel for Mac_ v16.71 (23031200).

Works all good using excel online.

Using `fetch` to make API calls.

This is affecting our production excel add-in.

Would be very helpful if you have any updates / solution on the same?

**Comment by b-g-d** [comment]
This still seems to be an issue in Word for Mac as of Jan, 2024! @JunielKatarn , any way to learn more about progress on this?

**Comment by JunielKatarn** [comment]
@b-g-d I am not involved in Word for Mac development. The issue I worked on for Custom Functions was specific to Excel for Windows.

@shaofengzhu can you please help us route this issue?

**Comment by timr11** [comment]
Hi, is there progress on this issue? I'm developing an Excel add-in on Mac Ventura 13.4 (22F66), Excel Version 16.84 (24041420), this is a major limitation. Would greatly appreciate this being addressed.

**Comment by BjornBirkelund** [comment]
This has also become a pressing issue for us, we are developing an excel add-in using react and are having our requests time out after 60s. Mac OS Sonoma 14.4.1 Version 16.85 (24051214) Microsoft 365 Subscription

**Comment by Alex-ray** [comment]
+1 on the preflight CORS request timeout configuration. This is a pressing issue for us as well.

**Comment by kbrilla** [comment]
For those following this thread — we've compiled a community open letter to Microsoft 
based on a systematic analysis of all 6,200+ issues in this repository. It documents 
patterns of silent breaking changes, auto-closed bugs, platform fragmentation, and the 
financial/reputation impact on ISVs and developers.

If any of this resonates with your experience, your voice would strengthen the message: 
[[link to open letter issue]](https://github.com/OfficeDev/office-js/issues/6513)

This is entirely voluntary — we just wanted to make sure affected developers are aware.

**Comment by kustonaut** [comment]
Hi All,
 
Picking this thread up as part of a broader effort to work through long-standing Office Add-ins issues, including the ones called out in the community open letter (#6513). Platform quality and trust here matter to us, and threads like this one are exactly where we want to start.

We re-tested this on the latest Excel for Windows and were not able to reproduce the 60-second timeout. On the shared runtime we tried both `GET` and `POST` calls to a local server that took longer than 60 seconds to respond, including a `POST` that goes through a CORS preflight. On the JavaScript-only runtime we tried a `GET` call against the same long-running server. In every case the custom function waited for the server and the result showed up correctly in the cell.

We have not retested on Mac yet which runs on a separate code path, and we know several of you originally hit this there.

Could those of you still seeing this please try again on a current build and let us know? If it still fails, a quick note on your platform, version, and what kind of call you are making would help us narrow it down. If it is working for you now, that is also useful as we will be able to close this out with a clear note for anyone landing here from search.

**Comment by kustonaut** [comment]
Since I have not received any response on an issue on current build I am closing this thread. 

We retested the original scenario on the latest Excel for Windows (shared and JavaScript-only runtimes, GET and POST including a CORS-preflighted POST against a >60s server) and the custom function returned the value correctly in every case — no 60-second timeout reproduced.

If anyone is still hitting this on a current build (especially on Excel for Mac, or specifically with CORS preflight on a POST), please open a new issue with:

Excel build (Excel → About Microsoft Excel) and host OS version
Runtime: shared or JavaScript-only
Request type: GET / POST / POST with preflight
Server response delay
Minimal repro (snippet or repo)

A fresh issue with current build numbers is much easier to action than reviving a 6-year-old thread. Please link back to this one for context.
